_posts/2023-10-19-large-scale-functions.md

```diff
@@ -0,0 +1,250 @@
+---
+title: "Scaling to 15000 functions and beyond"
+description: "Learn what changes we've made for scaling to tens of thousands of functions per installation"
+date: 2023-10-19
+categories:
+- functions
+- enterprise
+- saas
+- hosting
+dark_background: true
+author_staff_member: alex
+image: "/images/2023-10-scale-operator/background.jpg"
+hide_header_image: true
+---
+
+Learn what changes we made to help a customer scale to 15000 functions and beyond.
+
+## Introduction
+
+In this post I'll give an overview of what we learned spending a week investigating a customer issue with scaling beyond 3500 functions. Whilst navigating this issue, we also implemented several optimisations and built new tooling for testing Kubernetes at scale.
+
+If you've ever written software, which other people install and support, then you'll know how difficult and time-consuming it can be to debug and diagnose problems remotely. In this case it was no different, with our team spending over a week of R&D trying to reproduce the problem, pin-point the cause, and remediate it.
+
+**How many functions is a normal amount?**
+
+First of all, I should explain that we are only aware of 2-3 users running over 2000-3000 functions in production, so this is more of an edge case than the norm for OpenFaaS teams. It's not that we discourage large scale, or function hosting, it's just that OpenFaaS is popular with individual project teams who have a modest number of functions. So, teams that I've talked to over the years, tend to run with 2-5 dozen functions, with a number of others well below the 3000 mark.
+
+**What does it cost to test at scale?**
+
+In addition, with OpenFaaS running on top of Kubernetes, we have to provision a whole node for every 100-110 containers that are provisioned, including control-plane, service mesh and networking. So for 3000 functions, you need at least 30 nodes.
+
+We all know that clusters are slow to create on a platform like AWS EKS, and then adding nodes can take a good 3-5 minutes each. I did a calculation with the AWS cost estimator and to get to 3500 functions, you'd probably be looking at a spend of 1500 USD / mo in infrastructure costs alone.
+
+**How did we find the problem?**
+
+The problem was finally found after spending after building over a week of optimizations, and it was frustratingly obvious.
```

Comment:
This sentence has a weird structure. You intention was probably to write something like:

```suggestion
The problem was finally found after spending over a week building  optimizations, and it was frustratingly obvious.
```